include/extractor/guidance/mergable_road_detector.hpp

```diff
@@ -0,0 +1,149 @@
+#define OSRM_EXTRACTOR_GUIDANCE_MERGEABLE_ROADS
+
+#include "extractor/guidance/coordinate_extractor.hpp"
+#include "extractor/guidance/intersection.hpp"
+#include "extractor/guidance/intersection_generator.hpp"
+#include "extractor/query_node.hpp"
+#include "extractor/suffix_table.hpp"
+
+#include "util/coordinate_calculation.hpp"
+#include "util/name_table.hpp"
+#include "util/node_based_graph.hpp"
+#include "util/typedefs.hpp"
+
+#include <cstdint>
+#include <functional>
+#include <limits>
+#include <vector>
+
+namespace osrm
+{
+namespace extractor
+{
+namespace guidance
+{
+
+class MergableRoadDetector
+{
+  public:
+    // in case we have to change the mode we are operating on
+    using MergableRoadData = IntersectionShapeData;
+
+    MergableRoadDetector(const util::NodeBasedDynamicGraph &node_based_graph,
+                         const std::vector<QueryNode> &node_coordinates,
+                         const IntersectionGenerator &intersection_generator,
+                         const CoordinateExtractor &coordinate_extractor,
+                         const util::NameTable &name_table,
+                         const SuffixTable &street_name_suffix_table);
+
+    /*
+     * OSM ways tend to be modelled as separate ways for different directions. This is often due to
+     * small gras strips in the middle between the two directions or due to pedestrian islands at
+     * intersections.
+     *
+     * To reduce unnecessary information due to these artificial intersections (which are not
+     * actually perceived as such) we try and merge these for our internal representation to both
+     * get better perceived turn angles and get a better reprsentation of our intersections in
+     * general.
+     *
+     *         i   h                                             i,h
+     *         |   |                                              |
+     *         |   |                                              |
+     * b - - -   v   - - - g                                      |
+     *         > a <               is transformed into: b,c - - - a - - -  g,f
+     * c - - -   ^   - - - f                                      |
+     *         |   |                                              |
+     *         |   |                                              |
+     *         d   e                                             d,e
+     */
+    bool CanMergeRoad(const NodeID intersection_node,
+                      const MergableRoadData &lhs,
+                      const MergableRoadData &rhs) const;
+
+    /* check if a road cannot influence the merging of the other. This is necessary to prevent
+     * situations with more than two roads that could participate in a merge */
+    bool IsDistinctFrom(const MergableRoadData &lhs, const MergableRoadData &rhs) const;
+
+  private:
+    // check if two name ids can be seen as identical (in presence of refs/others)
+    bool NameIDsAreCompatible(const NameID lhs, const NameID rhs) const;
```

Comment:
This is the name change heuristic in both directions. Clarified in comment. It uses the name change heuristic. Also changed name to `HaveIdenticalNames`